alicloud/resource_alicloud_polardb_cluster.go

```diff
@@ -0,0 +1,447 @@
+		return resourceAlicloudPolarDBClusterRead(d, meta)
+	}
+
+	if d.HasChange("description") {
+		request := polardb.CreateModifyDBClusterDescriptionRequest()
+		request.RegionId = client.RegionId
+		request.DBClusterId = d.Id()
+		request.DBClusterDescription = d.Get("description").(string)
+
+		raw, err := client.WithPolarDBClient(func(polarDBClient *polardb.Client) (interface{}, error) {
+			return polarDBClient.ModifyDBClusterDescription(request)
+		})
+		if err != nil {
+			return WrapErrorf(err, DefaultErrorMsg, d.Id(), request.GetActionName(), AlibabaCloudSdkGoERROR)
+		}
+		addDebug(request.GetActionName(), raw, request.RpcRequest, request)
+		d.SetPartial("description")
+	}
+
+	if d.HasChange("security_ips") {
+		ipList := expandStringList(d.Get("security_ips").(*schema.Set).List())
+
+		ipstr := strings.Join(ipList[:], COMMA_SEPARATED)
+		// default disable connect from outside
+		if ipstr == "" {
+			ipstr = LOCAL_HOST_IP
+		}
+
+		if err := polarDBService.ModifyDBSecurityIps(d.Id(), ipstr); err != nil {
+			return WrapError(err)
+		}
+		d.SetPartial("security_ips")
+	}
+
+	d.Partial(false)
+	return resourceAlicloudPolarDBClusterRead(d, meta)
+}
+
+func resourceAlicloudPolarDBClusterRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	polarDBService := PolarDBService{client}
+
+	clusterAttribute, err := polarDBService.DescribePolarDBClusterAttribute(d.Id())
+	if err != nil {
+		if polarDBService.NotFoundCluster(err) {
+			d.SetId("")
+			return nil
+		}
+		return WrapError(err)
+	}
+
+	cluster, err := polarDBService.DescribePolarDBCluster(d.Id())
+	if err != nil {
+		if polarDBService.NotFoundCluster(err) {
+			d.SetId("")
+			return nil
+		}
+		return WrapError(err)
+	}
+
+	clusterAutoRenew, err := polarDBService.DescribePolarDBAutoRenewAttribute(d.Id())
+	if err != nil {
+		if polarDBService.NotFoundCluster(err) {
+			d.SetId("")
+			return nil
+		}
+		return WrapError(err)
+	}
+
+	ips, err := polarDBService.GetSecurityIps(d.Id())
+	if err != nil {
+		return WrapError(err)
+	}
+
+	d.Set("security_ips", ips)
+
+	d.Set("vswitch_id", clusterAttribute.VSwitchId)
+	d.Set("pay_type", clusterAttribute.PayType)
+	d.Set("id", clusterAttribute.DBClusterId)
+	d.Set("description", clusterAttribute.DBClusterDescription)
+	d.Set("db_type", clusterAttribute.DBType)
+	d.Set("db_version", clusterAttribute.DBVersion)
+	d.Set("maintain_time", clusterAttribute.MaintainTime)
+	d.Set("zone_ids", clusterAttribute.ZoneIds)
+	d.Set("cluster_charge_type", clusterAttribute.PayType)
+	d.Set("cluster_network_type", clusterAttribute.DBClusterNetworkType)
+	d.Set("db_node_class", cluster.DBNodeClass)
+	d.Set("auto_renew", clusterAutoRenew.AutoRenewEnabled)
+	period := 1
+	if clusterAutoRenew != nil {
+		period = clusterAutoRenew.Duration
+	}
+	if clusterAutoRenew != nil && clusterAutoRenew.PeriodUnit == string(Year) {
+		period = period * 12
+	}
+	d.Set("auto_renew_period", period)
+	d.Set("period", d.Get("period"))
```

Comment:
The set value can not come from d.get. If the API does not return it, please use createtime and endtime to compute it. Please refer to kvstore instance period.